// overrides/core.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getOverrides = getOverrides;
exports.getComponents = getComponents;

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var React = require("react");

exports.initializeOverrideWrappers = function initializeOverrideWrappers(defaultComponents, getOverridesProp) {
  // initialize component wrappers
  var components = {};

  var _loop = function _loop() {
    var name = _Object$keys[_i];

    components[name] = function (props) {
      var overrides = getOverridesProp() || {};
      var override = overrides[name];

      var _getOverrides = getOverrides(overrides[name], defaultComponents[name], props),
          _getOverrides2 = _slicedToArray(_getOverrides, 2),
          Component = _getOverrides2[0],
          mergedProps = _getOverrides2[1];

      return React.createElement(Component, mergedProps);
    };

    components[name].displayName = "".concat(name, "Overridable");
  };

  for (var _i = 0, _Object$keys = Object.keys(defaultComponents); _i < _Object$keys.length; _i++) {
    _loop();
  }

  return components;
};

function getOverrides(override, Component) {
  var props = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

  // component override shortcut:
  if (typeof override === "function" || typeof override === "string" || override instanceof React.Component) {
    Component = override;
  } else if (override) {
    var style = override.style,
        propsOverride = override.props,
        component = override.component,
        nested = _objectWithoutProperties(override, ["style", "props", "component"]); // assume at least one of the folllowing will override props, so clone


    props = _objectSpread({}, props); // component override:

    if (component) {
      Component = component;
    } // props override:


    if (propsOverride) {
      Object.assign(props, propsOverride);
    } // style override:


    if (style) {
      props.style = maybeMerge(props.style, typeof style === "function" ? style(props) : style); // TODO: support $style?
      // props.$style = style;
    } // nested overrides:


    if (Object.keys(nested).length > 0) {
      props.overrides = maybeMerge(props.overrides, nested);
    }
  }

  return [Component, props];
}

function getComponents(defaultComponents, overrides) {
  return Object.keys(defaultComponents).reduce(function (acc, name) {
    var _getOverrides3 = getOverrides(overrides[name], defaultComponents[name]),
        _getOverrides4 = _slicedToArray(_getOverrides3, 2),
        component = _getOverrides4[0],
        props = _getOverrides4[1];

    acc[name] = {
      component: component,
      props: props
    };
    return acc;
  }, {});
}

function maybeMerge(a, b) {
  return a && b ? _objectSpread({}, a, b) : a || b;
}
